File: operator/bridges/shared/test_bridge_paths.py
"""ADR-0008 Phase 8.1 per-subtask E2E for bridge_runtime_dir()."""
from __future__ import annotations

import os
import sys
import tempfile
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
import paths  # noqa: E402


def _clean_env():
    for key in list(os.environ.keys()):
        if key.startswith("CORVIN_BRIDGE") or key in (
            "CORVIN_HOME", "CORVIN_BRIDGES_HOME", "CORVIN_HOME",
        ):
            os.environ.pop(key, None)


def case_corvin_home_root() -> None:
    with tempfile.TemporaryDirectory() as tmp:
        _clean_env()
        os.environ["CORVIN_HOME"] = tmp
        assert paths.bridges_home() == Path(tmp) / "bridges"
        assert paths.bridge_channel_dir("discord") == Path(tmp) / "bridges" / "discord"


def case_kind_resolution() -> None:
    with tempfile.TemporaryDirectory() as tmp:
        _clean_env()
        os.environ["CORVIN_HOME"] = tmp
        for kind in ("inbox", "outbox", "processed", "attachments", "auth", "log"):
            assert paths.bridge_runtime_dir("telegram", kind) == \
                Path(tmp) / "bridges" / "telegram" / kind, kind
        # settings + root both return the channel dir
        for alias in ("settings", "root"):
            assert paths.bridge_runtime_dir("telegram", alias) == \
                Path(tmp) / "bridges" / "telegram", alias


def case_convenience_paths() -> None:
    with tempfile.TemporaryDirectory() as tmp:
        _clean_env()
        os.environ["CORVIN_HOME"] = tmp
        assert paths.bridge_settings_path("discord") == \
            Path(tmp) / "bridges" / "discord" / "settings.json"
        assert paths.bridge_log_path("email") == \
            Path(tmp) / "bridges" / "email" / "log" / "voice.log"


def case_root_env_override() -> None:
    with tempfile.TemporaryDirectory() as tmp:
        _clean_env()
        os.environ["CORVIN_HOME"] = "/should/not/be/used"
        os.environ["CORVIN_BRIDGES_HOME"] = tmp
        assert paths.bridge_runtime_dir("slack", "inbox") == \
            Path(tmp) / "slack" / "inbox"


def case_per_leaf_env_override() -> None:
    _clean_env()
    os.environ["CORVIN_HOME"] = "/should/not/be/used"
    os.environ["CORVIN_BRIDGE_DISCORD_INBOX"] = "/sandbox/inbox-override"
    assert paths.bridge_runtime_dir("discord", "inbox") == \
        Path("/sandbox/inbox-override")
    # other channel/kind not affected by this override
    assert paths.bridge_runtime_dir("discord", "outbox") != \
        Path("/sandbox/inbox-override")


def case_invalid_channel_rejected() -> None:
    _clean_env()
    for bad in ("Discord", "bad/channel", "", "a" * 33, "9digits", "with space"):
        try:
            paths.bridge_runtime_dir(bad, "inbox")
        except ValueError:
            continue
        raise AssertionError(f"should have rejected channel {bad!r}")


def case_invalid_kind_rejected() -> None:
    _clean_env()
    for bad in ("Bogus", "INBOX", "messages", "", "settings.json"):
        try:
            paths.bridge_runtime_dir("discord", bad)
        except ValueError:
            continue
        raise AssertionError(f"should have rejected kind {bad!r}")


def case_identity_only_no_fs_side_effects() -> None:
    with tempfile.TemporaryDirectory() as tmp:
        _clean_env()
        os.environ["CORVIN_HOME"] = tmp
        # Walk every channel × every queue kind. None of these should
        # have created directories — the migration helper is the sole
        # mkdir owner.
        for channel in ("telegram", "discord", "slack", "whatsapp", "email"):
            for kind in ("inbox", "outbox", "processed", "attachments", "auth", "log"):
                p = paths.bridge_runtime_dir(channel, kind)
                assert not p.exists(), f"resolver MUST NOT create {p}"
        assert not (Path(tmp) / "bridges").exists(), \
            "bridges_home MUST NOT be created"


def case_legacy_path_points_at_repo() -> None:
    _clean_env()
    legacy = paths.legacy_bridge_runtime_dir("email", "attachments")
    assert legacy is not None
    assert legacy.is_absolute()
    # the legacy path is the in-repo path
    assert "operator/bridges/email/attachments" in str(legacy)


def case_known_channels_all_accepted() -> None:
    _clean_env()
    with tempfile.TemporaryDirectory() as tmp:
        os.environ["CORVIN_HOME"] = tmp
        for ch in ("telegram", "discord", "slack", "whatsapp", "email", "shared"):
            assert paths.bridge_runtime_dir(ch, "inbox").parent.name == ch


CASES = [
    case_corvin_home_root,
    case_kind_resolution,
    case_convenience_paths,
    case_root_env_override,
    case_per_leaf_env_override,
    case_invalid_channel_rejected,
    case_invalid_kind_rejected,
    case_identity_only_no_fs_side_effects,
    case_legacy_path_points_at_repo,
    case_known_channels_all_accepted,
]


def main() -> int:
    failed = []
    for case in CASES:
        try:
            case()
            print(f"PASS {case.__name__}")
        except Exception as e:
            print(f"FAIL {case.__name__}: {e!r}")
            failed.append(case.__name__)
    if failed:
        print(f"\n{len(failed)} failure(s):", failed)
        return 1
    print(f"\nAll {len(CASES)} cases passed.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: operator/bridges/shared/js/test_bridge_paths.js
// ADR-0008 Phase 8.1 per-subtask E2E for bridgeRuntimeDir() in Node.
//
// Mirrors operator/bridges/shared/test_bridge_paths.py case-for-case
// so any future divergence between the Python and Node resolver shows up
// as a failing case here.

'use strict';

const fs = require('fs');
const os = require('os');
const path = require('path');

function cleanEnv() {
  for (const key of Object.keys(process.env)) {
    if (key.startsWith('CORVIN_BRIDGE') ||
        key === 'CORVIN_HOME' ||
        key === 'CORVIN_BRIDGES_HOME') {
      delete process.env[key];
    }
  }
}

function fresh() {
  // Resolver caches nothing — but re-require to be safe across cases.
  delete require.cache[require.resolve('./bridge_paths')];
  return require('./bridge_paths');
}

function mktmp() {
  return fs.mkdtempSync(path.join(os.tmpdir(), 'adr0008-'));
}

function assertEq(a, b, msg) {
  if (a !== b) throw new Error(`${msg || 'assertion'}: ${a} !== ${b}`);
}

function caseCorvinHomeRoot() {
  cleanEnv();
  const tmp = mktmp();
  process.env.CORVIN_HOME = tmp;
  const p = fresh();
  assertEq(p.bridgesHome(), path.join(tmp, 'bridges'));
  assertEq(p.bridgeChannelDir('discord'), path.join(tmp, 'bridges', 'discord'));
}

function caseKindResolution() {
  cleanEnv();
  const tmp = mktmp();
  process.env.CORVIN_HOME = tmp;
  const p = fresh();
  for (const kind of ['inbox', 'outbox', 'processed', 'attachments', 'auth', 'log']) {
    assertEq(
      p.bridgeRuntimeDir('telegram', kind),
      path.join(tmp, 'bridges', 'telegram', kind),
      kind,
    );
  }
  for (const alias of ['settings', 'root']) {
    assertEq(
      p.bridgeRuntimeDir('telegram', alias),
      path.join(tmp, 'bridges', 'telegram'),
      alias,
    );
  }
}

function caseConveniencePaths() {
  cleanEnv();
  const tmp = mktmp();
  process.env.CORVIN_HOME = tmp;
  const p = fresh();
  assertEq(p.bridgeSettingsPath('discord'),
    path.join(tmp, 'bridges', 'discord', 'settings.json'));
  assertEq(p.bridgeLogPath('email'),
    path.join(tmp, 'bridges', 'email', 'log', 'voice.log'));
}

function caseRootEnvOverride() {
  cleanEnv();
  const tmp = mktmp();
  process.env.CORVIN_HOME = '/should/not/be/used';
  process.env.CORVIN_BRIDGES_HOME = tmp;
  const p = fresh();
  assertEq(p.bridgeRuntimeDir('slack', 'inbox'),
    path.join(tmp, 'slack', 'inbox'));
}

function casePerLeafEnvOverride() {
  cleanEnv();
  process.env.CORVIN_HOME = '/should/not/be/used';
  process.env.CORVIN_BRIDGE_DISCORD_INBOX = '/sandbox/inbox-override';
  const p = fresh();
  assertEq(p.bridgeRuntimeDir('discord', 'inbox'), '/sandbox/inbox-override');
  if (p.bridgeRuntimeDir('discord', 'outbox') === '/sandbox/inbox-override') {
    throw new Error('per-leaf override leaked to a different kind');
  }
}

function caseInvalidChannelRejected() {
  cleanEnv();
  const p = fresh();
  for (const bad of ['Discord', 'bad/channel', '', 'a'.repeat(33), '9digits', 'with space']) {
    let rejected = false;
    try { p.bridgeRuntimeDir(bad, 'inbox'); } catch (e) { rejected = true; }
    if (!rejected) throw new Error(`should have rejected channel ${JSON.stringify(bad)}`);
  }
}

function caseInvalidKindRejected() {
  cleanEnv();
  const p = fresh();
  for (const bad of ['Bogus', 'INBOX', 'messages', '', 'settings.json']) {
    let rejected = false;
    try { p.bridgeRuntimeDir('discord', bad); } catch (e) { rejected = true; }
    if (!rejected) throw new Error(`should have rejected kind ${JSON.stringify(bad)}`);
  }
}

function caseIdentityOnlyNoFsSideEffects() {
  cleanEnv();
  const tmp = mktmp();
  process.env.CORVIN_HOME = tmp;
  const p = fresh();
  for (const channel of ['telegram', 'discord', 'slack', 'whatsapp', 'email']) {
    for (const kind of ['inbox', 'outbox', 'processed', 'attachments', 'auth', 'log']) {
      const resolved = p.bridgeRuntimeDir(channel, kind);
      if (fs.existsSync(resolved)) {
        throw new Error(`resolver MUST NOT create ${resolved}`);
      }
    }
  }
  if (fs.existsSync(path.join(tmp, 'bridges'))) {
    throw new Error('bridgesHome MUST NOT be created');
  }
}

function caseLegacyPathPointsAtRepo() {
  cleanEnv();
  const p = fresh();
  const legacy = p.legacyBridgeRuntimeDir('email', 'attachments');
  if (!legacy || !path.isAbsolute(legacy)) {
    throw new Error('legacy path must be absolute');
  }
  // Accept both the new operator/bridges layout and the legacy plugins/voice/bridges
  // path — the function falls back to whichever exists.
  const newPath = path.join('operator', 'bridges', 'email', 'attachments');
  const legacyPath = path.join('plugins', 'voice', 'bridges', 'email', 'attachments');
  if (!legacy.includes(newPath) && !legacy.includes(legacyPath)) {
    throw new Error(`legacy path should be in repo tree: ${legacy}`);
  }
}

function caseKnownChannelsAllAccepted() {
  cleanEnv();
  const tmp = mktmp();
  process.env.CORVIN_HOME = tmp;
  const p = fresh();
  for (const ch of ['telegram', 'discord', 'slack', 'whatsapp', 'email', 'shared']) {
    const resolved = p.bridgeRuntimeDir(ch, 'inbox');
    if (path.basename(path.dirname(resolved)) !== ch) {
      throw new Error(`channel resolution wrong for ${ch}: ${resolved}`);
    }
  }
}

const CASES = [
  caseCorvinHomeRoot,
  caseKindResolution,
  caseConveniencePaths,
  caseRootEnvOverride,
  casePerLeafEnvOverride,
  caseInvalidChannelRejected,
  caseInvalidKindRejected,
  caseIdentityOnlyNoFsSideEffects,
  caseLegacyPathPointsAtRepo,
  caseKnownChannelsAllAccepted,
];

let failed = 0;
for (const c of CASES) {
  try {
    c();
    console.log(`PASS ${c.name}`);
  } catch (e) {
    console.log(`FAIL ${c.name}: ${e.message}`);
    failed += 1;
  }
}
if (failed) {
  console.log(`\n${failed} failure(s).`);
  process.exit(1);
}
console.log(`\nAll ${CASES.length} cases passed.`);
